ci/tcli/tcli.go

```diff
@@ -0,0 +1,328 @@
+		"-it", "-d",
+	}
+	args = append(args, mountArgs(config.BindMounts)...)
+	args = append(args, config.Image)
+
+	cmd := exec.CommandContext(ctx, "docker", args...)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, xerrors.Errorf(
+			"failed to start testing container %q, (%s): %w",
+			config.Name, string(out), err)
+	}
+
+	return &ContainerRunner{
+		name: config.Name,
+		ctx:  ctx,
+	}, nil
+}
+
+// Close kills and removes the command execution testing container
+func (r *ContainerRunner) Close() error {
+	cmd := exec.CommandContext(r.ctx,
+		"sh", "-c", strings.Join([]string{
+			"docker", "kill", r.name, "&&",
+			"docker", "rm", r.name,
+		}, " "))
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return xerrors.Errorf(
+			"failed to stop testing container %q, (%s): %w",
+			r.name, string(out), err)
+	}
+	return nil
+}
+
+// Run executes the given command in the runtime container with reasonable defaults.
+// "command" is executed in a shell as an argument to "sh -c".
+func (r *ContainerRunner) Run(ctx context.Context, command string) *Assertable {
+	cmd := exec.CommandContext(ctx,
+		"docker", "exec", "-i", r.name,
+		"sh", "-c", command,
+	)
+
+	return &Assertable{
+		cmd:   cmd,
+		tname: command,
+	}
+}
+
+// RunCmd lifts the given *exec.Cmd into the runtime container
+func (r *ContainerRunner) RunCmd(cmd *exec.Cmd) *Assertable {
+	path, _ := exec.LookPath("docker")
+	cmd.Path = path
+	command := strings.Join(cmd.Args, " ")
+	cmd.Args = append([]string{"docker", "exec", "-i", r.name}, cmd.Args...)
+
+	return &Assertable{
+		cmd:   cmd,
+		tname: command,
+	}
+}
+
+// HostRunner executes command tests on the host, outside of a container
+type HostRunner struct{}
+
+// Run executes the given command on the host.
+// "command" is executed in a shell as an argument to "sh -c".
+func (r *HostRunner) Run(ctx context.Context, command string) *Assertable {
+	cmd := exec.CommandContext(ctx, "sh", "-c", command)
+
+	return &Assertable{
+		cmd:   cmd,
+		tname: command,
+	}
+}
+
+// RunCmd executes the given *exec.Cmd on the host
+func (r *HostRunner) RunCmd(cmd *exec.Cmd) *Assertable {
+	return &Assertable{
+		cmd:   cmd,
+		tname: strings.Join(cmd.Args, " "),
+	}
+}
+
+// Close is a noop for HostRunner
+func (r *HostRunner) Close() error {
+	return nil
+}
+
+// Assertable describes an initialized command ready to be run and asserted against
+type Assertable struct {
+	cmd   *exec.Cmd
+	tname string
+}
+
+// Assert runs the Assertable and
+func (a Assertable) Assert(t *testing.T, option ...Assertion) {
+	slog.Helper()
+	var (
+		stdout bytes.Buffer
+		stderr bytes.Buffer
+		result CommandResult
+	)
+
+	a.cmd.Stdout = &stdout
+	a.cmd.Stderr = &stderr
+
+	start := time.Now()
+	err := a.cmd.Run()
+	result.Duration = time.Since(start)
+
+	if exitErr, ok := err.(*exec.ExitError); ok {
+		result.ExitCode = exitErr.ExitCode()
+	} else if err != nil {
+		// TODO: handle this case better
+		result.ExitCode = -1
```

Comment:
Could we fail the test here and log the error?

Reply:
Yeah, that seems like the best option for now.